src/Storages/StorageMergeTree.cpp

```diff
@@ -1543,28 +1545,61 @@ void StorageMergeTree::replacePartitionFrom(const StoragePtr & source_table, con
 
     Stopwatch watch;
     MergeTreeData & src_data = checkStructureAndGetMergeTreeData(source_table, source_metadata_snapshot, my_metadata_snapshot);
-    String partition_id = getPartitionIDFromQuery(partition, local_context);
+    String partition_id = src_data.getPartitionIDFromQuery(partition, local_context);
 
     DataPartsVector src_parts = src_data.getVisibleDataPartsVectorInPartition(local_context, partition_id);
     MutableDataPartsVector dst_parts;
     std::vector<scope_guard> dst_parts_locks;
 
     static const String TMP_PREFIX = "tmp_replace_from_";
 
-    for (const DataPartPtr & src_part : src_parts)
+    auto query_to_string = [] (const ASTPtr & ast)
+    {
+        return ast ? queryToString(ast) : "";
+    };
+
+    const auto is_partition_exp_different = query_to_string(my_metadata_snapshot->getPartitionKeyAST()) != query_to_string(source_metadata_snapshot->getPartitionKeyAST());
```

Comment:
How would that work if both `my_metadata_snapshot->getPartitionKeyAST()` and `source_metadata_snapshot->getPartitionKeyAST()` are `nullptr`?

Reply:
I guess the partition exp would be considered to be the same. The very same check is already present in [here](https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeData.cpp#L6141), I just moved it over here.